File: lemur-libs/lemur-testkit/src/main/java/io/github/malczuuu/lemur/testkit/kafka/KafkaTestAutoConfiguration.java
package io.github.malczuuu.lemur.testkit.kafka;

import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.context.ApplicationListener;
import org.springframework.context.annotation.Bean;
import org.springframework.context.event.ContextRefreshedEvent;
import org.springframework.kafka.config.KafkaListenerEndpointRegistry;

/**
 * Auto-configuration that waits for all Kafka listeners to be assigned before the application
 * context is fully refreshed. This ensures that any tests relying on Kafka listeners will not start
 * until the listeners are ready to receive messages, preventing race conditions and ensuring
 * reliable test execution.
 */
@AutoConfiguration
@ConditionalOnBean(KafkaListenerEndpointRegistry.class)
public final class KafkaTestAutoConfiguration {

  @Bean
  ApplicationListener<ContextRefreshedEvent> kafkaAssignmentAwaiter(
      KafkaListenerEndpointRegistry registry) {
    return _ -> KafkaTestUtils.awaitAssignment(registry);
  }
}

File: lemur-libs/lemur-testkit/src/main/kotlin/io/github/malczuuu/lemur/testkit/kafka/KafkaTestAutoConfiguration.kt
package io.github.malczuuu.lemur.testkit.kafka

import org.awaitility.Awaitility.await
import org.springframework.boot.autoconfigure.AutoConfiguration
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.context.ApplicationListener
import org.springframework.context.annotation.Bean
import org.springframework.context.event.ContextRefreshedEvent
import org.springframework.kafka.config.KafkaListenerEndpointRegistry
import java.time.Duration

@AutoConfiguration
@ConditionalOnBean(KafkaListenerEndpointRegistry::class)
class KafkaTestAutoConfiguration {

    @Bean
    fun kafkaAssignmentAwaiter(registry: KafkaListenerEndpointRegistry): ApplicationListener<ContextRefreshedEvent> =
        ApplicationListener { registry.awaitAssignment() }

    private fun KafkaListenerEndpointRegistry.awaitAssignment() {
        await()
            .atMost(Duration.ofSeconds(30))
            .pollInterval(Duration.ofMillis(200))
            .until { hasAllPartitionsAssigned() }
    }

    private fun KafkaListenerEndpointRegistry.hasAllPartitionsAssigned(): Boolean =
        allListenerContainers.all { container ->
            val partitions = container.assignedPartitions
            container.isRunning && !partitions.isNullOrEmpty()
        }
}
